Add python keyword search quickstart (#2106)
elastic/docs-content: solutions/search/get-started/keyword-search-python.md

These clients offer full API support for indexing, searching, and cluster management.
They are optimized for performance and kept up to date with {{es}} releases, ensuring compatibility and security.

## Prerequisites

This quickstart does not require previous knowledge of {{es}} but assumes a basic familiarity with Python development.
To follow the steps, you must have a recent version of a Python interpreter.

## Create an {{es-serverless}} project

Create an [{{es-serverless}}](/solutions/search/serverless-elasticsearch-get-started.md) general purpose project.
To add the sample data, you must have a `developer` or `admin` predefined role or an equivalent custom role.
To learn about role-based access control, go to [](/deploy-manage/users-roles/cluster-or-deployment-auth/user-roles.md).

## Create an index

An index is a collection of documents uniquely identified by a name or an alias.
To create an index, go to **{{es}} > Home**, select keyword search, and follow the guided workflow.

To enable your client to talk to your project, you must also create an API key.
Click **Create API Key** and use the default values, which are sufficient for this quickstart.

:::{tip}

> For more information about indices and API keys, go to [](/manage-data/data-store/index-basics.md) and [](/deploy-manage/api-keys/serverless-project-api-keys.md).

Reviewer: the `[ ](url)` syntax automatically adds the page/section title as the text of the link :) 
(second "Do" here: https://elastic.github.io/docs-builder/syntax/quick-ref/#links)

---
description: An introduction to building an Elasticsearch query in Python. 
applies_to:
  serverless: all
products:
  - id: elasticsearch
---
# Build your first search query with Python

{{es}} provides a range of search techniques, starting with BM25, the industry standard for textual search.
It provides official clients for multiple programming languages, including Python, Rust, Java, JavaScript, and others. 
These clients offer full API support for indexing, searching, and cluster management.

> They are optimized for performance and kept up to date with {{es}} releases, ensuring compatibility and security.

Author: Thanks! Fixed in c547fabf159343da50cd9d8e2cdc82e9600ff314

In this quickstart guide, you will index a couple documents and query them using Python.

> By the end of this guide, you’ll have learned how to connect a backend application to {{es}} to answer your queries.

Author: Great advice! Fixed in c547fabf159343da50cd9d8e2cdc82e9600ff314

> 

Reviewer: here's a screenshot that shows my suggestions for the first 2 paragraphs
<img width="816" height="208" alt="Screenshot 2025-07-16 at 5 28 39 PM" src="https://github.com/user-attachments/assets/fa600064-c246-43fd-93e9-3a8b801c7879" />

## Prerequisites

- If you're using [{{es-serverless}}](/solutions/search/serverless-elasticsearch-get-started.md), create a general purpose project. To add the sample data, you must have a `developer` or `admin` predefined role or an equivalent custom role.

> <!--

Author: Yeah, I'll likely not get it re-added before this needs to go live, so for now i've removed the bullet in c547fabf159343da50cd9d8e2cdc82e9600ff314

If you're using [{{ech}}](/deploy-manage/deploy/elastic-cloud/cloud-hosted.md) or [running {{es}} locally](/solutions/search/run-elasticsearch-locally.md), start {{es}} and {{kib}}. To add the sample data, log in with a user that has the `superuser` built-in role.
-->
  
To learn about role-based access control, check out [](/deploy-manage/users-roles/cluster-or-deployment-auth/user-roles.md).

## Create an index

An index is a collection of documents uniquely identified by a name or an alias.
Go to **{{es}} > Home**, select keyword search, and follow the guided index workflow.
<!--
Click **Create a TBD index**.
- If you're using {{es-serverless}}...
- If you're using {{ech}} or running {{es}} locally, go to **{{es}} > Home** and click **Create API index**. Select the semantic search workflow.
-->

You've created your first index!
Next, create an API key so your application can talk to {{es}}.
<!--
TBD: Describe how to create the key
-->
:::{tip}
For an introduction to the concept of indices, check out [](/manage-data/data-store/index-basics.md).
:::

> ## Install an {{es}} client

Reviewer: Should we go ahead and say "the Python client" ?

Author: Sure! Added in c547fabf159343da50cd9d8e2cdc82e9600ff314

Select your preferred language in the keyword search workflow. For this example, leverage Python.

![Client installation step in the index management workflow](https://images.contentstack.io/v3/assets/bltefdd0b53724fa2ce/bltbf810f73fd4082fb/67c21c06304ea9790b82ee4d/screenshot-my-index.png)

In your terminal, install the {{es}} client using `pip`:

```py
pip install elasticsearch
```

Copy your API key from the top right corner and add it to the client’s configuration alongside the project URL.

```py
from elasticsearch import Elasticsearch

client = Elasticsearch(
    "https://my-project-bff300.es.us-east-1.aws.elastic.cloud:443",
    api_key="YOUR-API-KEY"
)

index_name = "my-index"
```

## Create field mappings

> At this stage, you can define the mappings for your index, including a single text field — named "text".

Reviewer: I mostly did this as a workaround for regional differences re: quotation marks and punctuation 😁

    },
    {
        "text": "Rocky Mountain National Park  is one of the most popular national parks in the United States. It receives over 4.5 million visitors annually, and is known for its mountainous terrain, including Longs Peak, which is the highest peak in the park. The park is home to a variety of wildlife, including elk, mule deer, moose, and bighorn sheep. The park is also home to a variety of ecosystems, including montane, subalpine, and alpine tundra. The park is a popular destination for hiking, camping, and wildlife viewing, and is a UNESCO World Heritage Site."
    }
]

bulk_response = helpers.bulk(client, docs, index=index_name)
print(bulk_response)
```

## Explore the data

You should be able to see the documents in {{es}}!

![Viewing data in the index management workflow](https://images.contentstack.io/v3/assets/bltefdd0b53724fa2ce/blt0ac36402cde2a645/67d0a443b8764e72b9e8e1f3/view_docs_in_elasticsearch.png)
<!--
To familiarize yourself with this data set, open [Discover](/explore-analyze/discover.md) from the navigation menu or the global search field.
-->

## Test keyword search

> Create a new script (for instance `search.py`), which defines a query and runs the following search request:

Reviewer: ```suggestion
Create a new script (for instance, `search.py`) that defines a query and runs the following search request:
```

        	| WHERE MATCH(text, "yosemite")
        	| LIMIT 5
    	""",
	format="csv"
)

print(response)
```

## Analyze the results

Check your result:

```txt
"Yosemite National Park is a United States National Park, covering over 750,000 acres of land in California. A UNESCO World Heritage Site, the park is best known for its granite cliffs, waterfalls and giant sequoia trees. Yosemite hosts over four million visitors in most years, with a peak of five million visitors in 2016. The park is home to a diverse range of wildlife, including mule deer, black bears, and the endangered Sierra Nevada bighorn sheep. The park has 1,200 square miles of wilderness, and is a popular destination for rock climbers, with over 3,000 feet of vertical granite to climb. Its most famous and cliff is the El Capitan, a 3,000 feet monolith along its tallest face."
Now you are ready to use the client to query Elasticsearch from any Python backend like Flask, Django, etc. Check out the Elasticsearch Python Client documentation to explore further
```

<!--
When you finish your tests and no longer need the sample data set, delete the index:

```console
DELETE /semantic-index
```
-->

## Next steps

> Thanks for taking the time to learn how to build an application on top of {{es}}.

Reviewer: ```suggestion
This quickstart covered the basics of working with {{es}}.
```

> For a deeper dive, check out the following resources:

Reviewer: ```suggestion
For a deeper dive, refer to the following resources:
```
